### DISCUSSION TITLE
Use of Go programing language for FDO implementation

### DISCUSSION THREAD

**Discussion by rftemple** [secure-device-onboard/pri-fidoiot#681] [Ideas]
The purpose of this discussion is to collect feedback from adopters on the potential use of the Go programming language for implementing the FDO specification. 
The LF Edge FDO project currently uses Java and C as the primary programming languages for implementing the FIDO specification.   Java is primarily used for the server-side implementation of the protocol while C is primarily used for the client (device) side implementation of the protocol.   The FDO team has been investigating implementing the FDO specification using the Go programming language.   There are some advantages that this investigation has found.   First, Go can be used for both client and server-side implementation.   Also, the specification can be implemented entirely using core Go packages with very limited need for any 3rd party packages.   On the device-side Go compiles to a single static native binary that can be used without 3rd party libraries.   On the server-side Go can compile to very small microservice containers.   
The proposal would be to create a new repository under https://github.com/fido-device-onboard called go-fdo.   Under this repository would the go implementation be worked on.  Both the Java, C, and Go implementations would co-exist for a time with the expectation that the ecosystem would eventually move over to the Go implementation.

**Comment by naphelps** [comment]
Hello,

I am a maintainer with the LF Edge project Open Horizon. We have an API layer over top of the FDO provided ownership services demo written in Go. [https://github.com/open-horizon/FDO-support](https://github.com/open-horizon/FDO-support)

**Comment by ben-krieger** [comment]
The go-fdo library has been released, but we'd still love feedback and what else it can and should do! For example, it introduces a new idea of plugin service info modules that speak a line-based protocol over stdin/stdout so they can be implemented in any language and updated separately from the client/server executables.
